metasynth/distribution/faker.py

```diff
@@ -59,3 +64,109 @@ def _param_schema(cls):
+
+    This distribution detects the language and generates sentences using
+    the Faker package. The average number of sentences and words per item
+    are detected using regexes (with the lingua package).
+
+    Parameters
+    ----------
+    locale: str
+        Locale used for the faker package.
+    avg_sentences:
+        Average number of sentences (punctuation marks) per (non-NA) row,
+        if None do not make sentences.
+    avg_words:
+        Average number of words per (non-NA) row.
+    """
+
+    def __init__(self, locale: str, avg_sentences: Optional[float], avg_words: int):
+        self.locale: str = locale
+        self.avg_sentences = avg_sentences
+        self.avg_words = avg_words
+        self.fake = Faker(locale=self.locale)
+
+    @classmethod
+    def _fit(cls, values):
+        """Select the appropriate faker function and locale."""
+        lang_str = cls.detect_language(values)
+        if lang_str is None:
+            return cls.default_distribution()
+
+        try:
+            Faker(lang_str)
+        except AttributeError:
+            lang_str = "EN"
+
+        all_text = "\n".join(values)
+        n_non_empty = (values != "").sum()
+        n_punctuation = len(list(PUNCTUATION.finditer(all_text)))
+        n_words = len(list(LETTERS.finditer(all_text)))
+        if n_punctuation < n_non_empty//3:
+            avg_sentence = None
+        else:
+            avg_sentence = n_punctuation/len(values)
+        avg_words = n_words/len(values)
+        return cls(lang_str, avg_sentence, avg_words)
+
+    @classmethod
+    def detect_language(cls, values: Iterable) -> Optional[str]:
+        """Detect the language of some text.
+
+        Parameters
+        ----------
+        values:
+            Values to detect the language of (usually polars dataframe).
+
+        Returns
+        -------
+        language:
+            Two letter ISO code to represent the language, or None if it could not be determined.
+        """
+        detector = LanguageDetectorBuilder.from_all_languages().with_low_accuracy_mode().build()
+        lang = detector.detect_language_of("\n".join(values))
+        if lang is None:
+            return None
+        return str(lang.iso_code_639_1).rsplit(".", maxsplit=1)[-1]
+
+    def draw(self):
+        if self.avg_sentences is None:
+            n_words = max(1, poisson(self.avg_words).rvs())
+            sentence = self.fake.sentence(n_words)
+            return sentence[:-1]
+
+        n_sentences = max(1, poisson(self.avg_sentences).rvs())
+        avg_words_per_sent = max(1, self.avg_words/max(1, self.avg_sentences))
+        n_words = max(1, poisson(avg_words_per_sent).rvs())
+        return " ".join(self.fake.sentence(nb_words=n_words) for _ in range(n_sentences))
+
+    def information_criterion(self, values) -> float:
+        series = self._to_series(values)
+        lang = self.detect_language(series)
+        if lang is None:
+            return 9999999
+        return -1
```

Comment:
Is this what makes the choice between regex and unstructured? Can we make it more explicit somehow? This logic is now hidden in the detect_language function right?

Reply:
Yes, if lingua detects a language -> UnstructuredText, otherwise  -> Regex. The logic is in the lingua package, so I don't really know how to make it more explicit? I can add a comment?